### ISSUE TITLE
BPE Tokenizer: Multiple newlines doesn't merge into a single token

### ISSUE THREAD

**Issue by Lyrcaxis** [ggerganov/llama.cpp#6809]
So, I found out that `\n\n` if appended by a character tokenizes as `['\n',\n']` (`[198, 198]`) instead of `['\n\n']` (`[271]`).
(I'm using Llama3 for this example, but this extends to other models as well)

Here's an example prompt:
```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

You're Psy, user's assistant, and a master of concise replies.<|eot_id|><|start_header_id|>user<|end_header_id|>

Write a short poem<|eot_id|><|start_header_id|>assistant<|end_header_id|>


```
And the tokenized text:
![image](https://github.com/ggerganov/llama.cpp/assets/32474602/b78fc581-59ea-408c-bc27-a14dc13bb7ad)

If I switch the template to use `\n\n\n\n` (`1038`) it tokenizes as `['\n\n\n', '\n']` (`[1432, 198]`):
![image](https://github.com/ggerganov/llama.cpp/assets/32474602/fdcd6093-81f6-4de2-8f85-d57f6def8d46)

(Note: I know there've been efforts in making special tokens render, but rn I understand they don't have a textual representation, so you can ignore tokens like 128000, 128006 and 128007 in the sequences above)

In C# I patch the issue like so:
```cs
var tokensCount = NativeApi.llama_tokenize(model, bytesPtr, bytes.Length, tokensPtr, tokenBuffer.Length, add_bos, special);
var list = new List<LLamaToken>();
for (int i = 0; i < tokensCount; i++) { // Hack: ['\n','\n'] --> ['\n\n']
    if (tokenBuffer[i] == 198 && tokenBuffer[i + 1] == 198) { list.Add(271); i++; }
    else { list.Add(tokenBuffer[i]); }
}
return list.ToArray();
```
(ignoring all \n merges except the `\n\n` which is common for the template)

**Comment by MarcusDunn** [comment]
I'm also running into this. It seems degrades performance for llama-3-instruct. (Hackily replacing two new line with the single token improves performance anecdotally) 

I'd imagine there are other cases where the tokenization is not as greedy as possible - unsure how this would affect model performance though.

**Comment by bullno1** [comment]
The `bpe_gpt2_preprocess` split the string `\n\nword` in a bit of a strange way: `\n`, `\nword`.

See: https://github.com/ggerganov/llama.cpp/pull/5613

**Comment by LostRuins** [comment]
I am encountering a similar issue. For me, the model likes to generate token `.\n` (id=627) at the end of the sentence. However, when retokenizing the string subsequently I instead get two disjoint tokens `. ` (id=13), and `\n` (id=198) 

Same thing with various other tokens like `.\n\n` (id=382)
Something is really broken with the merging behavior related to newlines.

**Comment by ggerganov** [comment]
Does anyone know what regex is used by LLaMA 3 to preprocess the text?

In `llama.cpp` we currently implement just this:

https://github.com/ggerganov/llama.cpp/blob/aa750c1ede6232c91de890a14a7731d6daa2bc8e/llama.cpp#L12128-L12135

My guess is that we need to update it with whatever is used for LLaMA 3

**Comment by LostRuins** [comment]
I'm not sure how accurate this is, but here is [a possible reference](https://belladoreai.github.io/llama3-tokenizer-js/example-demo/build/) which appears to at least merge `\n\n` correctly: https://raw.githubusercontent.com/belladoreai/llama3-tokenizer-js/master/llama-tokenizer.js

However, it's not just implemented as a clean regex but appears to have some additional processing too.

**Comment by MarcusDunn** [comment]
> Does anyone know what regex is used by LLaMA 3 to preprocess the text?

Is this what you'd be looking for?

https://github.com/meta-llama/llama3/blob/af6eedf7042fb51d00b2b26d8ef1ceaab73e1670/llama/tokenizer.py#L47

**Comment by ggerganov** [comment]
The issue should be fixed with #6920

**Comment by Lyrcaxis** [comment]
Awesome!
